src/Symfony/Component/DependencyInjection/Compiler/ResolveBindingsPass.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\LazyProxy\ProxyHelper;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Guilhem Niot <guilhem.niot@gmail.com>
+ */
+class ResolveBindingsPass extends AbstractRecursivePass
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition || !$bindings = $value->getBindings()) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        foreach ($bindings as $key => $binding) {
+            if (isset($key[0]) && '$' === $key[0]) {
+                continue;
+            }
+
+            if (null !== $binding && !$binding instanceof Reference && !$binding instanceof Definition) {
+                throw new InvalidArgumentException(sprintf('Invalid value for binding type "%s" for service "%s": expected null, an instance of %s or an instance of %s, %s given.', $key, (string) $this->currentId, Reference::class, Definition::class, gettype($binding)));
+            }
+
+            if (!class_exists($key) && !interface_exists($key)) {
```

Comment:
but thinking more about this, you told me you would prefer throwing when a binding is never used.
we *can* to that: we just need to move the deep clone that is done in FileLoader to ResolveDefinitionInheritancePass!
I'd prefer this over this class_exists check myself (because it makes things context-sensitive)